--- sdplane/l2fwd.orig.c	2024-05-12 23:43:18.411508892 +0900
+++ sdplane/l2fwd.c	2024-05-12 23:40:07.992573354 +0900
@@ -38,10 +38,20 @@
 #include <rte_mbuf.h>
 #include <rte_string_fns.h>
 
+#include <zcmdsh/debug.h>
+#include <zcmdsh/command.h>
+#include "sdplane.h"
+
+/* override rte_exit() so that the whole process is not broken. */
+#define rte_exit(x, ...) \
+  do { printf (__VA_ARGS__); return -1; } while (0)
+#define rte_warn(x, ...) \
+  do { printf (__VA_ARGS__); } while (0)
+
 static volatile bool force_quit;
 
-/* MAC updating enabled by default */
-static int mac_updating = 1;
+/* MAC updating disabled for sdplane. */
+static int mac_updating = 0;
 
 /* Ports set in promiscuous mode off by default. */
 static int promiscuous_on;
@@ -64,10 +74,10 @@
 static struct rte_ether_addr l2fwd_ports_eth_addr[RTE_MAX_ETHPORTS];
 
 /* mask of enabled ports */
-static uint32_t l2fwd_enabled_port_mask = 0;
+uint32_t l2fwd_enabled_port_mask = 0;
 
 /* list of enabled ports */
-static uint32_t l2fwd_dst_ports[RTE_MAX_ETHPORTS];
+uint32_t l2fwd_dst_ports[RTE_MAX_ETHPORTS];
 
 struct port_pair_params {
 #define NUM_PORTS	2
@@ -78,7 +88,7 @@
 static struct port_pair_params *port_pair_params;
 static uint16_t nb_port_pair_params;
 
-static unsigned int l2fwd_rx_queue_per_lcore = 1;
+unsigned int l2fwd_rx_queue_per_lcore = 1;
 
 #define MAX_RX_QUEUE_PER_LCORE 16
 #define MAX_TX_QUEUE_PER_PORT 16
@@ -113,7 +123,7 @@
 static uint64_t timer_period = 10; /* default period is 10 seconds */
 
 /* Print out statistics on packets dropped */
-static void
+void
 print_stats(void)
 {
 	uint64_t total_packets_dropped, total_packets_tx, total_packets_rx;
@@ -176,6 +186,36 @@
 	rte_ether_addr_copy(&l2fwd_ports_eth_addr[dest_portid], &eth->src_addr);
 }
 
+struct rte_ring *tap_ring_by_lcore[RTE_MAX_LCORE];
+__thread struct rte_ring *thread_ring_to_tap;
+
+static void
+l2fwd_copy_to_tap_ring (struct rte_mbuf *m, unsigned portid)
+{
+  struct rte_mbuf *c;
+  uint32_t pkt_len;
+  uint16_t data_len;
+  char *m_start, *c_start;
+  pkt_len = rte_pktmbuf_pkt_len (m);
+  data_len = rte_pktmbuf_data_len (m);
+  if (FLAG_CHECK (debug_config, DEBUG_SDPLANE_WIRETAP))
+    printf ("%s: m: %p (%d/%d) from port: %d\n",
+            __func__, m, data_len, pkt_len, portid);
+#if 1
+  c = rte_pktmbuf_copy (m, m->pool, 0, UINT32_MAX);
+#elif 0
+  c = rte_pktmbuf_clone (m, m->pool);
+#else
+  c = rte_pktmbuf_alloc (m->pool);
+  data_len = rte_pktmbuf_data_len (m);
+  rte_pktmbuf_append (c, data_len);
+  m_start = rte_pktmbuf_mtod (m, char *);
+  c_start = rte_pktmbuf_mtod (c, char *);
+  memcpy (c_start, m_start, data_len);
+#endif
+  rte_ring_enqueue (thread_ring_to_tap, c);
+}
+
 /* Simple forward. 8< */
 static void
 l2fwd_simple_forward(struct rte_mbuf *m, unsigned portid)
@@ -232,7 +272,7 @@
 
 	}
 
-	while (!force_quit) {
+	while (! force_quit && ! force_stop[lcore_id]) {
 
 		/* Drains TX queue in its main loop. 8< */
 		cur_tsc = rte_rdtsc();
@@ -291,6 +331,7 @@
 			for (j = 0; j < nb_rx; j++) {
 				m = pkts_burst[j];
 				rte_prefetch0(rte_pktmbuf_mtod(m, void *));
+				l2fwd_copy_to_tap_ring (m, portid);
 				l2fwd_simple_forward(m, portid);
 			}
 		}
@@ -298,7 +339,7 @@
 	}
 }
 
-static int
+int
 l2fwd_launch_one_lcore(__rte_unused void *dummy)
 {
 	l2fwd_main_loop();
@@ -647,7 +688,7 @@
 }
 
 int
-main(int argc, char **argv)
+l2fwd_init (int argc, char **argv)
 {
 	struct lcore_queue_conf *qconf;
 	int ret;
@@ -659,6 +700,7 @@
 	unsigned int nb_lcores = 0;
 	unsigned int nb_mbufs;
 
+#if 0
 	/* Init EAL. 8< */
 	ret = rte_eal_init(argc, argv);
 	if (ret < 0)
@@ -677,6 +719,7 @@
 	/* >8 End of init EAL. */
 
 	printf("MAC updating %s\n", mac_updating ? "enabled" : "disabled");
+#endif
 
 	/* convert to number of cycles */
 	timer_period *= rte_get_timer_hz();
@@ -745,6 +788,8 @@
 
 		/* get the lcore_id for this port */
 		while (rte_lcore_is_enabled(rx_lcore_id) == 0 ||
+		       lcore_workers[rx_lcore_id].func !=
+		       l2fwd_launch_one_lcore ||
 		       lcore_queue_conf[rx_lcore_id].n_rx_port ==
 		       l2fwd_rx_queue_per_lcore) {
 			rx_lcore_id++;
@@ -900,6 +945,7 @@
 			"All available ports are disabled. Please set portmask.\n");
 	}
 
+#if 0
 	check_all_ports_link_status(l2fwd_enabled_port_mask);
 
 	ret = 0;
@@ -927,6 +973,7 @@
 	/* clean up the EAL */
 	rte_eal_cleanup();
 	printf("Bye...\n");
+#endif
 
 	return ret;
 }
